app/src/main/java/org/oppia/app/home/topiclist/TopicListAdapter.kt

```diff
@@ -0,0 +1,126 @@
+package org.oppia.app.home.topiclist
+
+import android.view.LayoutInflater
+import android.view.ViewGroup
+import androidx.databinding.DataBindingUtil
+import androidx.databinding.ViewDataBinding
+import androidx.recyclerview.widget.RecyclerView
+import androidx.databinding.library.baseAdapters.BR
+import org.oppia.app.R
+import org.oppia.app.databinding.PromotedStoryCardBinding
+import org.oppia.app.databinding.TopicSummaryViewBinding
+import org.oppia.app.databinding.WelcomeBinding
+import org.oppia.app.home.UserAppHistoryViewModel
+import org.oppia.app.model.TopicSummary
+
+private const val VIEW_TYPE_WELCOME_MESSAGE = 1
+private const val VIEW_TYPE_PROMOTED_STORY = 2
+private const val VIEW_TYPE_TOPIC_LIST = 3
+
+/** Adapter to inflate different items/views inside [RecyclerView]. The itemList consists of various ViewModels. */
+class TopicListAdapter(
+  private val itemList: MutableList<Any>,
+  private val topicSummaryClickListener: TopicSummaryClickListener
+) :
+  RecyclerView.Adapter<RecyclerView.ViewHolder>() {
+
+  override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
+    return when (viewType) {
+      // TODO(#249): Generalize this binding to make adding future items easier.
+      VIEW_TYPE_WELCOME_MESSAGE -> {
+        val inflater = LayoutInflater.from(parent.context)
+        val binding =
+          DataBindingUtil.inflate<WelcomeBinding>(
+            inflater,
+            R.layout.welcome,
+            parent,
+            /* attachToParent= */false
+          )
+        WelcomeViewHolder(binding)
+      }
+      VIEW_TYPE_PROMOTED_STORY -> {
+        val inflater = LayoutInflater.from(parent.context)
+        val binding =
+          DataBindingUtil.inflate<PromotedStoryCardBinding>(
+            inflater,
+            R.layout.promoted_story_card,
+            parent,
+            /* attachToParent= */false
+          )
+        PromotedStoryViewHolder(binding)
+      }
+      VIEW_TYPE_TOPIC_LIST -> {
+        val inflater = LayoutInflater.from(parent.context)
+        val binding =
+          DataBindingUtil.inflate<TopicSummaryViewBinding>(
+            inflater,
+            R.layout.topic_summary_view,
+            parent,
+            /* attachToParent= */false
+          )
+        TopicListViewHolder(binding)
+      }
+      else -> throw IllegalArgumentException("Invalid view type") as Throwable
```

Comment:
Would be nice to include the view type designated as invalid in the error string to provide context if it's ever thrown.

Reply:
Updated error to `IllegalArgumentException("Invalid view type: $viewType")`